add delete all comments button

poncem91 opened the pull request:
Commit adds button that allows user delete all the store comments permanently.

poncem91 committed:
add delete all comments button

Commit adds button that allows user delete all the store comments permanently.

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures' into Mafe_AdditionalFeatures_DeleteComments

merging recent changes from base branch

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures' into Mafe_AdditionalFeatures_DeleteComments

add recent commit from base branch

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures' into Mafe_AdditionalFeatures_DeleteComments

apply changes from commit in base branch

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -129,4 +129,12 @@ function getComments(maxComments) {
+function deleteComments() {
+  const request = new Request("/delete-comments", {method: 'POST'});
+  fetch(request).then(() => {
+    getComments(0);
```
I think you're calling `getComments(0)` in your response callback here so that the comments section of the page gets updated, which is good. However, I think that asking for zero comments could backfire; imagine a bug is introduced such that the backend doesn't delete all comments, but still returns an OK response. The frontend would turn around and fetch zero comments, and it would look to the user like there was a success. However, the remaining comments would reappear upon page reload.

I think the hypothetical I described is unlikely, but still I think it would be better to fetch comments with the same limit as normal, in case something weird happens; it's better for the user to notice a problem right away.

poncem91 replied:
Ah, I hadn't thought of that aspect! I'll change this to avoid this potential issue, thanks!

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServletDelete.java:
```diff
@@ -0,0 +1,52 @@
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that allows users to add comments and returns comment history  */
+@WebServlet("/delete-comments")
```
It's not introduced in the walkthroughs AFAIK, but there's a better HTTP method for this: DELETE. Since we're still operating on comments here, we can move the handler to the same class as the other comment handlers and reuse the /comments URL with the DELETE method to perform this function.

poncem91 replied:
Great, I'll look into this! I'll try to incorporate the single comment delete feature into this as well.

peap replied:
Sounds good; looks like you've done that in #17.

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServletDelete.java:
```diff
@@ -0,0 +1,52 @@
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+        datastore.delete(entity.getKey());
```
Just mentioning something that came up in another review.

In general, I like to make the database do things for us in as few requests as possible, so a bulk delete of matching records is better than many single-record deletes. But that only works if the database supports it, and the Datastore docs make it look like you have to use Dataflow for that. I don't think we need to bring in that dependency just for this.

poncem91 replied:
I would've thought that there was a relatively simple way of batch deleting everything as well, but found the same mention of using Dataflow in the docs. I'll keep this in mind for the future though!

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures' into Mafe_AdditionalFeatures_DeleteComments

merge changes done in base branch

poncem91 committed:
Merge branch 'master' into Mafe_AdditionalFeatures_DeleteComments

merge base branch changes

poncem91 committed:
use DELETE HTTP method to delete comments

Commit uses the DELETE HTTP request method instead of the POST method to delete the comments and in turn is able to reuse the /comments URL with a doDelete() function instead.
